rust/arrow/src/buffer/mod.rs

```diff
@@ -0,0 +1,26 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! This module contains two main structs: [Buffer] and [MutableBuffer]. A buffer represents
+//! a contiguous memory region that can be shared via `offsets`.
+
+mod immutable;
+pub use immutable::*;
+mod mutable;
+pub use mutable::*;
+mod ops;
+pub(super) use ops::*;
```

Comment:
Wouldn't this change be breaking to users who use the functions in `ops`, as they've been crate-public?

Reply:
FWIW it might not be possible bit I think in general making `Buffer` and `MutableBuffer` etc less exposed is a good direction -- they become implementation details that can then be more freely changed without affecting users (e.g. it would make things like https://github.com/jorgecarleitao/arrow2/pull/1 easier) . 

In other words, making ops crate public might be a *good* thing (even if it is a potentially breaking one)

I realize the cat may already be out of the bag there, but I think being more deliberate about what interfaces are public and what are private would help developers and users going forward

Reply:
> making ops crate public might be a _good_ thing

My view's that if someone's manipulating bits in a general way that's not related to Arrow, they shouldn't be using these `ops`. So I agree with not making them public.

After looking at https://doc.rust-lang.org/reference/visibility-and-privacy.html, I see that I used the wrong terminology.
As now making `arrow::buffer::buffer_bin_or` inaccessible to arrow's consumers, this is a semantic change.

@jorgecarleitao WDYT about returning it to `pub use ops::*`, and then we can address which of our laundry we hang to the public, in future?

Reply:
Sorry @nevi-me  -- I meant "making ops crate *not* public might be a good thing"

Reply:
How about we just make another follow on PR that makes `ops` pub again?

Reply:
First, this PR is not merged yet; I just mistakenly pushed it to master and reverted it 😞 Anyways...

AFAI can tell, `buffer_bin_or` is currently not public; `pub(super)` just makes it public to the arrow crate in this context (as `super` of `buffer.rs` is `arrow`).

In master, I can't make the below compile:

```rust
extern crate arrow;

use arrow::buffer::Buffer;
use arrow::error::Result;
use arrow::buffer::buffer_bin_or;

fn main() -> Result<()> {
    let a = Buffer::from_slice_ref(&[1u32, 2u32]);

    let b = (&a | &a)?;
    Ok(())
}
```

However, `And` and `Or` are indeed public and this PR makes them private. I just pushed a change to revert that by exposing the implementations in `mod.rs`.